SPEC

```diff
@@ -113,6 +113,25 @@ be implemented by the server.
                         fatal(message, &block)
 <tt>rack.multipart.buffer_size</tt>:: An Integer hint to the multipart parser as to what chunk size to use for reads and writes.
 <tt>rack.multipart.tempfile_factory</tt>:: An object responding to #call with two arguments, the filename and content_type given for the multipart form field, and returning an IO-like object that responds to #<< and optionally #rewind. This factory will be used to instantiate the tempfile for each multipart form file upload field, rather than the default class of Tempfile.
+
+Servers that choose to support WebSockets and SSE connections should follow these additional environment specifications:
+<tt>rack.upgrade?</tt>:: Is nil or missing if no connection upgrade is possible,
+                         :websocket for a WebSocket upgrade, or :sse for an SSE upgrade.
+<tt>rack.upgrade</tt>:: Is used to pass a handler instance or class back to the server for WebSocket and SSE connections.
+                     The handler MAY implement any or all of the following callbacks:
+                      on_open()                        # called when the connection is complete
+                      on_message(message)              # called when a WebSocket message is received by server. <tt>message</tt> will be UTF-8 encoded for text messages and Binary encoded for binary messages.
+                      on_shutdown()                    # may be called before a connection is closed due to server shutdown.
+                      on_close()                       # called when the connection is closed
+                      on_drained                       # may be called when the number of pending writes drops to zero.
+                     The object will be extended by the server to include:
```

Comment:
This will break method caches.  I'd prefer a solution that doesn't lean on runtime method extension.  Maybe something like this:

```ruby
class MyWSEObject
  def initialize io
    @io = io
  end

  def on_open
    @io.write("neat")
  end
end

app = lambda do |env|
  env['rack.upgrade'] = MyWSEObject.new(env['rack.wse.object'])
end
```

Reply:
```
class MyWSEObject
  def on_open(io)
    io.write("neat")
  end
end

app = lambda do |env|
  env['rack.upgrade'] = MyWSEObject.new
end
```

would be simpler

Reply:
That would be fine too. 😄

Reply:
So, the next question, is it a good idea to write into the request `env` to essentially send a response? To me, that's confusing. Perhaps at best it could be something like

```ruby
if upgrade = env['rack.upgrade']
  return upgrade.call(MyWSEObject.new)
end

return [400, {}, []]
```

Reply:
TBH it doesn't really matter to me.  Unfortunately though, `env` is really the only way that middleware can communicate.  So writing to `env` is the only way that one middleware could indicate to the next "I'm responding by writing to a websocket".  If the intention is that any "callable" (middleware / app) that sets the WSE object must not call anything else in the middleware chain, then that needs to be explicitly stated in SPEC.

Reply:
I understand the need to maintain the method caches, especially as we move towards JIT compilation.

As far as iodine is concerned, I am totally flexible on all points as long as the negative impact on performance isn't too high.

On the other hand, please consider a number of good reasons to prefer the `extend` approach over the `on_open(io)`:

1. It's likely that the method caches will break only the first time a callback object is returned.

   Assuming the same object class is used for all callback objects, the `extend` operations becomes a no-op, keeping the method caches intact (since the class was already extended).

2. Using an `io` object adds an obvious requirement for the `io` object to be stored somewhere by the receiver (otherwise, `write` could never get called).

   This makes the `on_open(io)` method a requirement rather than optional and it adds a required use-pattern (`def on_open(io); @io=io; end`).

   IMHO, although not always the case, this is a strong indication of a design flaw (requiring the user to complete the design).

3. For C extensions, there's double the amount of objects per connection. For Ruby implementations it's an increase of 150% in object count (I implemented in C, so my Ruby estimates are theoretical assuming 3 objects: raw IO, wrapper IO and callback objects).

   This puts an extra burden on the memory management system and hastens the (inevitable?) memory fragmentation issue that many Ruby servers suffer from.

4. As a side note, this might make future extensions harder, since future gem authors wouldn't be sure how to access the `io` object from within a connection.

I noticed that @evanphx 's experience is the opposite of mine.

While I found reduced performance and higher complexity when using an `io` object, @evanphx reports improved performance when using this approach.

Reply:
As for the `.call`, I find it confusing.

Let's assume I want to upgrade to SSE (or WebSockets) as well as set a cookie.

When using `call` (besides the extra Proc object), it's unclear to me when the `call` is actually performed. Is the upgrade immediate? Are the headers in the final response are sent?

Passing an object is both cheaper (no Proc object) and, I feel, simpler.

As for `[200, {}, MyObject]`, I think this breaks existing servers, so I'm pretty sure it wouldn't work as smoothly.

Reply:
> As for [200, {}, MyObject], I think this breaks existing servers, so I'm pretty sure it wouldn't work as smoothly.

As long as MyObject responds to `#each` as a fallback, it won't break existing servers.

Reply:
> > For C extensions, there's double the amount of objects per connection.

> Why is this?

@tenderlove , I assume you know that the C server doesn't need an IO object, it uses a simple integer to keep track of IO.

This means that a C extension just needs to keep the callback object itself and add a secret variable to that object with the socket number.

This secret variable isn't possible in Ruby, but C extensions are allowed to create them. Since that variable is a Number, it's very small (it doesn't create an Object in the ObjectSpace realm and it's immutable).

This means a single object in C (the callback object). If we add the IO object, it's two objects - a 200% increase in object count.

In Ruby, we will have the callback handler, the raw IO object (both pre-existing) and an IO wrapper, a 150% increase in object count.

Reply:
> This means a single object in C (the callback object). If we add the IO object, it's two objects - a 200% increase in object count.

What % increase is that in an actual app?

Reply:
> What % increase is that in an actual app?

That really depends on the app's use-case.

A micro services app running a WebSocket API will get hit harder than an HTTP app with the occasional WebSocket upload manager.

I think the issue is more pronounced because we are talking about long term objects.

Short term object count doesn't effect memory fragmentation as much as the fact that there's a long-term object blocking a memory "block" / "arena".

The more long-term objects are created (and eventually destroyed), the faster the fragmentation.

Having said that, even a 200% increase in object count is probably better than the solutions we currently employ when implementing WebSockets using `hijack`.

I won't make an issue out of it. I'll do what everyone thinks best, but I know in very practical terms (having coded both implementations in C) that `extend` is cheaper both in memory and in the number of method tree traversals required.